### ISSUE TITLE
Bug: [@typescript-eslint/no-explicit-any] any type should be allowed as generic constraint

### ISSUE THREAD

**Issue by overshom** [typescript-eslint/typescript-eslint#9449]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&showAST=ts&fileType=.tsx&code=MYewdgzgLgBA5gUzAgTgS2AdRQQwA4CqYA1mCAO5gwC8MAPAGJUIAeUSAJhDABQB0AnCjgQAXDACuJMpQDaAXQCUNAHyTpFMCp4AzMOKaKDVamoDeAKBgwUCKBJRUe-QcIjLTMPS75CRi5RxuJgBuCwBfC1BIWERkdCxcPABBMABPGnomGFZ2MC5eAV83cRx0hQ81MrTtPWMjGGzPS2tbe0deHz93VS8wLrcAmCDGsDDIiwB6ACoLQFByGAAlOwcqKDS8BBg0MB1UWw4YPBQQTZQAGzTRKPBoeBAQDgBJXdQkYARxHkqYMAkAWwARqgLNNJjcYvdHi89rYwB9MnFUBhsPgiKRNM4fgAWABMijCU1mgBlyGAAFQ2CAAysB0HhYABRFAnFDXZLCAFIWAgHQwdabGAAcl0D3E0HQYDgPzMXlFMHFOzgIRggKE4j%2BQNQyvCgu23DIsCCEDQcDAOEB5y2UBARyEOH%2BdlQMB5fMpQoGInEUgxciUvW9MjAgr4VnJlO4Lrwdod7BQ3EFOgeurKh0F3WTtm28JA-yjUDQFoQIes1gpAsFAc0urQ%2BpAhogxtN5stfJt-K2goVkuDPFxAGZsQBWRSg8HRO6Ak7EJAwt7wra0JEJVGEDSUZyJkBiqASqW9HgyzcAGhVapgg5g4QChJm8yWKw67azsIORxOZ0u13HsB0aBYCGeV44Q%2BL5N23XdpVDawwPlHdFTCEtT1ZX4AWBFBxlHCE7l-f9ANhd4F3gJBkUSfBUjSDc5S7PdPAPWUQBPVVkIvK8CQsIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge3oA9jGzKdAIZNC6KImjQO0SODABfEAqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
const genericWrapUnknown = <Fn extends (...args: unknown[]) => unknown>(fn: Fn): Fn => {
  return ((...args) => fn(...args)) as Fn;
}
const genericWrapAny = <Fn extends (...args: any[]) => any>(fn: Fn): Fn => {
  return ((...args) => fn(...args)) as Fn;
}

/*
✅ Return type inferred properly:
const goodInference: () => number
*/
const goodInference = genericWrapUnknown(() => 42);

/*
❌ TypeScript Error:
Argument of type '(foo: string) => { foo: string; bar: number; }' is not assignable to parameter of type '(...args: unknown[]) => unknown'.
  Types of parameters 'foo' and 'args' are incompatible.
    Type 'unknown' is not assignable to type 'string'.(2345)
*/
const brokenInference = genericWrapUnknown((foo: string) => ({ foo, bar: 5 }));

/*
✅ Return type inferred properly:
const fixedInference: (foo: string) => {
    foo: string;
    bar: number;
}
*/
const fixedInference = genericWrapAny((foo: string) => ({ foo, bar: 5 }));
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  "rules": {
    "@typescript-eslint/no-explicit-any": "error"
  }
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

I expect that `genericWrapAny` would have zero eslint errors about use of `any` types because they are used in generic constraints which makes them safe to use because actual type will be inferred from function arguments safely.

### Actual Result

`genericWrapAny` reports errors about use of `any` types inside generic constraints.

But if you look at `fixedInference` inferred type - there is no single `any` type there meaning type inferred properly and fully type safe to use.

### Additional Info

The issue of type inference arises in TypeScript `strict` mode for compiler options.
TypeScript version is latest - 5.5.2.

Without `strict` it can work fine with just `unknown` constraints, but `strict` mode requires to use `any` for proper type inference.

Also, it seems the issue is only about type inference for `args` of a generic function.
If we have `unknown` type as return type compiler can infer fine.

> Maybe it is a bug of TypeScript compiler itself, as I would expect `unknown` type to work for described cases as good as `any` type.

Anyway - `any` type is fine to use for generic constraints.
So please adjust the rule for such cases.

**Comment by Josh-Cena** [comment]
Hi, `(...args: unknown[]) => unknown` does not represent an arbitrary function. You need to use `(...args: never) => unknown` instead, which gives you everything you want while avoiding `any`.
